src/pal/src/numa/numa.cpp

```diff
@@ -867,3 +867,114 @@ VirtualAllocExNuma(
 
     return result;
 }
+
+/*++
+Function:
+  SetThreadIdealProcessorEx
+
+See MSDN doc.
+--*/
+BOOL
+PALAPI
+SetThreadIdealProcessorEx(
+  IN HANDLE hThread,
+  IN PPROCESSOR_NUMBER lpIdealProcessor,
+  OUT PPROCESSOR_NUMBER lpPreviousIdealProcessor)
+{
+    PERF_ENTRY(SetThreadIdealProcessorEx);
+    ENTRY("SetThreadIdealProcessorEx(hThread=%p, lpIdealProcessor=%p)\n", hThread, lpIdealProcessor);
+
+    CPalThread *pCurrentThread = InternalGetCurrentThread();
+    CPalThread *pTargetThread = NULL;
+    IPalObject *pTargetThreadObject = NULL;
+
+    PAL_ERROR palErr =
+        InternalGetThreadDataFromHandle(pCurrentThread, hThread,
+                                        0, // THREAD_SET_CONTEXT
+                                        &pTargetThread, &pTargetThreadObject);
+
+    if (NO_ERROR != palErr)
+    {
+        ERROR("Unable to obtain thread data for handle %p (error %x)!\n", hThread,
+              palErr);
+        return 0;
+    }
+
+    pthread_t thread = pTargetThread->GetPThreadSelf();
+
+#if HAVE_PTHREAD_GETAFFINITY_NP
+    if (lpPreviousIdealProcessor != NULL)
+    {
+        cpu_set_t prevCpuSet;
+        CPU_ZERO(&prevCpuSet);
+        DWORD prevCpu = GetCurrentProcessorNumber();
+
+        int st = pthread_getaffinity_np(thread, sizeof(cpu_set_t), &prevCpuSet);
+
+        if (st == 0)
+        {
+            for (int i = 0; i < g_possibleCpuCount; i++)
+            {
+                if (CPU_ISSET(i, &prevCpuSet))
+                {
+                    prevCpu = i;
+                    break;
+                }
+            }
+        }
+
+        _ASSERTE(prevCpu < g_possibleCpuCount);
+        lpPreviousIdealProcessor->Group = g_cpuToAffinity[prevCpu].Group;
```

Comment:
The lpPreviousIdealProcessor should be updated after we read the lpIdealProcessor values. As it is now, if the same pointer is passed in both these args, we would overwrite the lpIdealProcessor with the previous one and then apply the new one, which means nothing would change.

Reply:
Also, it would be good to set the Reserved member to 0.